uvicorn/server.py

```diff
@@ -283,13 +298,38 @@ async def shutdown(self, sockets: Optional[List[socket.socket]] = None) -> None:
         if not self.force_exit:
             await self.lifespan.shutdown()
 
-    def install_signal_handlers(self) -> None:
+    @contextlib.contextmanager
+    def _capture_exit_signals(self) -> Iterator[None]:
         if threading.current_thread() is not threading.main_thread():
             # Signals can only be listened to from the main thread.
-            return
+            yield
+        else:
+            original_handlers = {sig: signal.getsignal(sig) for sig in HANDLED_SIGNALS}
+            self.install_signal_handlers()
+            yield
+            for sig, handler in original_handlers.items():
+                signal.signal(sig, handler)
+            # if we did gracefully shut down due to a signal, try to
+            # trigger the expected behaviour for this signal now
+            if self._delayed_signal is not None:
```

Comment:
Can you help me understand a bit better why this is needed, and the scenario on which this is triggered?

If I got it right, this is triggered on `SIGINT` or `SIGTERM`, due to `handle_exit`, and it's triggered after the server is already down. But... Why do we need to call the original handlers? Why the `os.kill` below is not enough?

Reply:
The scenario is as you understood.

Calling the original signal handler directly allows avoiding the non-portability of `os.kill`, i.e. it works the same on Windows and others.  
Would you prefer the code to unconditionally use `os.kill` on non-Windows? That would probably be more in line with what people are used to.

Reply:
I've changed the order so that the common UNIX case is handled first. The workarounds are only tried on Windows-like platform.